pennylane/_device.py

```diff
@@ -439,7 +439,7 @@ def sample(self, observable, wires, par, n=None):
         raise NotImplementedError("Returning samples from QNodes not currently supported by {}".format(self.short_name))
 
     def probability(self):
-        """Return the full state probability from the last run of the device.
+        """Return the full state probability of each computational basis state from the last run of the device.
```

Comment:
I feel like this is still unclear. Why only support probabilities of computational basis states? Why from the last run of the device? (might fit better as something in `qml.measure`, e.g., `expval`, `var`, `sample`, `probability`, in which case, we might want to support arbitrary measurement POVMs

Reply:
Oh I see what you mean! The intention here is slightly different. We've had feedback from quite a few users (including @AroosaIjaz) that have asked for a standardized API to access 'raw' results from the device independently of the QNode. For example, after they have run a QNode, to be able to access:

* `device.state()` to return the statevector (for devices that support it)

* `dev.probability()` to return (marginal) probabilities

* `dev.samples()` to return raw samples generated by the device (for devices that support it)

These are quantities that are already calculated and stored by most of the plugins, there is just no unified API for them, so it differs between each plugin (for example, the SF and Forest plugins uses `dev.state`, default.qubit uses `dev._state`, the Qiskit and Forest plugins already have `dev.probability`, but with slightly different signatures, etc.). So this is mostly just unifying and standardizing it.

The main use-case would be users and developers being easily able to test and debug PennyLane code, for example,

```python
@qml.qnode(dev)
def circuit(x):
    qml.RX(x, wires=0)
    return expval(qml.PauliY(0))

assert circuit(np.pi/2) == -1
assert dev.probabilities() == np.array([0.5, 0.5])
```

and not as a main interface for machine learning with probability values.

Separate to this is the idea of supporting probability as a first class return value from QNodes. I'm not sure we need a new `probability()` function akin to `expval`, `var`, `sample`, mostly because probability outcomes can always be expressed as expectation values. All we would need to do is add new classes representing projectors as well as POVM elements, i.e.,

```python
@qml.qnode(dev)
def circuit(x):
    qml.RX(x, wires=0)
    return expval(qml.StateProjector(0)) # probability of measuring sigma_z = 1
```

^ but perhaps we need a better syntax, so that users can define their _own_ projectors/POVM elements to be used with the `expval` function.

Reply:
Understood. But I think the docstring of `probability` is still unclear. What if the user did not measure the computational basis states? Will this information still be available "from the last run of the device"?

Reply:
> What if the user did not measure the computational basis states? Will this information still be available "from the last run of the device"?

Yes. This method is more of a method to query the internal 'state' of the device between runs, and is independent of the QNode construction. Similar to how `device.state` already works. 

Note that this isn't 'user facing' similarly to how the QNode class methods aren't 'user facing', so we should not advertise them in the main documentation. But this may be useful for advanced users and developers.